core/src/main/java/com/datastax/oss/driver/internal/core/loadbalancing/BasicLoadBalancingPolicy.java

```diff
@@ -323,10 +323,7 @@ protected NodeDistanceEvaluator createNodeDistanceEvaluator(
           @Override
           protected Object[] computeNodes() {
             Object[] dcs = liveNodes.dcs().toArray();
-            if (dcs.length <= 1) {
-              return EMPTY_NODES;
-            }
-            Object[] remoteNodes = new Object[(dcs.length - 1) * maxNodesPerRemoteDc];
+            Object[] remoteNodes = new Object[dcs.length * maxNodesPerRemoteDc];
```

Comment:
Hey @Schnikonos , again my apologies for the delay in getting back to you on this.

This does fix the problem, but unfortunately we lose the benefit of the early return.  We're now relying on the !dc.equals(localDc) check in the for loop below before we can exit in the no-remote DC case.  Put differently we have to loop through the list of DCs twice in order to exit, once when constructing the array of DC names and again when looping through those DCs to add nodes.

I'm wondering if we can't improve on that by looping through DCs once before we get to the early return:

```java
         @Override
          protected Object[] computeNodes() {
            Set<String> remoteDcs =
                liveNodes.dcs().stream()
                    .filter(Predicates.not(Predicates.equalTo(localDc)))
                    .collect(Collectors.toSet());
            if (remoteDcs.isEmpty()) {
              return EMPTY_NODES;
            }
            Object[] remoteNodes = new Object[remoteDcs.size() * maxNodesPerRemoteDc];
            int remoteNodesLength = 0;
            for (String dc : remoteDcs) {
              Object[] remoteNodesInDc = liveNodes.dc(dc).toArray();
...
```

This works, but if we've come this far why not embrace the streaming API more completely?

```java
          @Override
          protected Object[] computeNodes() {
            Set<String> dcs = liveNodes.dcs();
            Object[] remoteNodes =
                dcs.stream()
                    .filter(Predicates.not(Predicates.equalTo(localDc)))
                    .flatMap(dc -> liveNodes.dc(dc).stream().limit(maxNodesPerRemoteDc))
                    .collect(Collectors.toList())
                    .toArray();

            int remoteNodesLength = remoteNodes.length;
            if (remoteNodesLength == 0) {
              return EMPTY_NODES;
            }
            shuffleHead(remoteNodes, remoteNodesLength);
            return remoteNodes;
          }
```

This feels like it might be the cleanest choice as it avoids the creation of the intermediate objects while also simplifying the logic considerably.

What do you think?